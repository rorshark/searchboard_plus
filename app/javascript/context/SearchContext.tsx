import React from 'react'
import { useSearch, SearchAPI, SearchParams } from '../hooks/useSearch'

export const defaultSearchApi: SearchAPI = {
  error: false,
  loading: true,
  customers: [],
  search: '',
  setSearch: (_) => {},
  companies: [],
  selectedCompany: null,
  setSelectedCompany: (_) => {},
  fetchCustomers: async () => {}
}

export const SearchContext: React.Context<SearchAPI> = React.createContext(defaultSearchApi)

export const SearchProvider: React.FC<{
  initialSearchParams: SearchParams
}> = ({ initialSearchParams, children }) => {
  const search = useSearch(initialSearchParams)

  return (
    <SearchContext.Provider value={search}>
      {children}
    </SearchContext.Provider>
  )
}
