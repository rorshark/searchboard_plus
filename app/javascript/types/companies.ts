export type Company = {
  companyName: string
}
