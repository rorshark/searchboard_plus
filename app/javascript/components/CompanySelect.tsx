import React, { useContext, useEffect } from 'react'
import { SearchContext } from '../context/SearchContext'

const CompanySelect: React.FC = () => {
  const { 
    companies, 
    fetchCompanies, 
    selectedCompany, 
    setSelectedCompany 
  } = useContext(SearchContext)

  const handleSelect = (evt: React.ChangeEvent<HTMLSelectElement>) => {
    const companyName = evt.target.value

    if (!companyName) {
      setSelectedCompany('')
      return
    }

    const company = companies.find((c) => c.companyName === companyName)

    if (!company) throw new Error(`${companyName} not found`)

    setSelectedCompany(company.companyName)
  }

  const isSelected = (name: string) => {
    return selectedCompany === name
  }

  useEffect(() => {
    fetchCompanies()
  }, [])

  return (
    <label>
      Company Name
      <select 
        name="company" 
        data-testid="company-select" 
        onChange={handleSelect}>
        <option value="">Please choose an option</option>

        {companies.map(({ companyName }, index) => (
          <option 
            selected={isSelected(companyName)} 
            data-testid={`company-${index}`} 
            key={index} 
            value={companyName}>
            {companyName}
          </option>
        ))}
      </select>
    </label>
  )
}

export default CompanySelect
