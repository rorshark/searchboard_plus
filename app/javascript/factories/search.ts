import { Factory } from 'fishery'
import { SearchAPI, SearchParams } from '../hooks/useSearch'
import { defaultSearchApi } from '../context/SearchContext'

export const searchApi = Factory.define<SearchAPI>(() => defaultSearchApi)

export const searchParams = Factory.define<SearchParams>(() => ({
  search: '',
  filter_by_company_name: ''
}))
