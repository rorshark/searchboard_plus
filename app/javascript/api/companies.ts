import axios from 'axios'
import { Company } from '../types/companies'

export type CompaniesResponse = {
  companies: Company[]
}

export const fetchCompanies = async (): Promise<CompaniesResponse> => {
  const { data } = await axios.get('/api/companies', {
    headers: {
      'Content-Type': 'application/json',
      'Key-Inflection': 'camel'
    }
  })

  return data
}
