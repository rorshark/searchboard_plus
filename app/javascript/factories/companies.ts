import { Factory } from 'fishery'
import Faker from 'faker'
import { Company } from '../types/companies'

export const company = Factory.define<Company>(() => ({
  companyName: Faker.company.companyName()
}))
