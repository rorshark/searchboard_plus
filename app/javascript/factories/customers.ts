import { Factory } from 'fishery'
import Faker from 'faker'
import { Customer } from '../types/customers'
import { company } from './companies'

export const customer = Factory.define<Customer>(() => ({
  firstName: Faker.name.firstName(),
  lastName: Faker.name.lastName(),
  company: company.build()
}));
