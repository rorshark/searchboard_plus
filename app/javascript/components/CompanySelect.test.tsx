import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { SearchContext } from '../context/SearchContext'
import { company } from '../factories/companies'
import { searchApi } from '../factories/search'
import CompanySelect from './CompanySelect'

test('it renders a dropdown list of companies', () => {
  const context = searchApi.build({
    companies: company.buildList(1, {
      companyName: 'New Relic'
    })
  })

  render(
    <SearchContext.Provider value={context}>
      <CompanySelect />
    </SearchContext.Provider>
  )

  const element = screen.getByTestId('company-select')

  expect(element).toMatchSnapshot()
})

test('selecting a company', () => {
  const setSelectedCompany = jest.fn()

  const context = searchApi.build({
    setSelectedCompany,
    companies: company.buildList(1, {
      companyName: 'New Relic'
    })
  })

  render(
    <SearchContext.Provider value={context}>
      <CompanySelect />
    </SearchContext.Provider>
  )

  const element = screen.getByTestId('company-select')

  fireEvent.change(element, { target: { value: 'New Relic' } })

  expect(setSelectedCompany.mock.calls[0][0]).toEqual('New Relic')
})
