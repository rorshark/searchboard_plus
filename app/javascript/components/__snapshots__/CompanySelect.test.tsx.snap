// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a dropdown list of companies 1`] = `
<input
  aria-hidden="true"
  class="MuiSelect-nativeInput"
  data-testid="company-select"
  tabindex="-1"
  value=""
/>
`;
